### ISSUE TITLE
GDScript: Make print use var2str for basic variables and auto-separation

### ISSUE THREAD

**Issue by xix-xeaon** [godotengine/godot-proposals#465]
**Describe the project you are working on:**
Any.

**Describe the problem or limitation you are having in your project:**
**print** really only exists for debugging purposes but `print(1, 1.0, "1")` gives `111` which is very confusing.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
Makes it much quicker to debug, and less confusing.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
`print(1, 1.0, "1")` now instead gives `1, 1.0, "1"` which is very clear and easy.

**Describe implementation detail for your proposal (in code), if possible:**
Implementation is simple. Single string argument and thing like Nodes etc should not use var2str. The only question would be to support different separators but I don't think it's necessary and since GDScript still doesn't support named arguments it wouldn't fit well.

I've already implemented it myself but I think it's a good candidate for someone who's never contributed before.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It will be used often and by everyone.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It is a core feature that everyone benefits from all the time.

**Comment by KoBeWi** [comment]
Are you aware of `prints` and `printt`?

**Comment by xix-xeaon** [comment]
Yes, in my opinion, prints is a pretty much useless functions since a simple space is not useful to clearly delineate between strings (which may have space in them) and sometimes not even between numbers, it can get confusing easily. printt might be useful sometimes, but I certainly prefer commas since that's how you write code so I've never actually used that one either, but most importantly neither function uses str2var.

I'd add str2var to all of them, but I don't think most people even know about them. I only found them when I patched the print function. I wouldn't mind if they were removed even - I had already mostly forgotten about them already. I just remembered to make this proposal today since the print function should probably be changed for 4.0.

**Comment by Calinou** [comment]
Many people use multiple `print()` arguments as a substitute for string concatenation, how would this play with the existing situation?

**Comment by xix-xeaon** [comment]
@Calinou This does break compatibility (although it's a minor thing). It would require people (who use it the way you say) to change from `print(a, b, c)` to `print(a+b+c)` and use `str(a)` to concatenate non-strings to get the exact same behavior. I would say using + is a better way of doing it although having to use str for some variables is inconvenient.

We also have format strings with % for pretty printing. I'm not against a dump function however, although I don't think I'd want each variable on it's own line most of the time (you "run out" of lines so quickly). Would dump also write the name of the variable? That would certainly be convenient like `dump(a, b, c)` gives `int a = 1; float b = 1.0; str c = "1";` or something. And maybe no arguments prints out all variables organized by scope or something - in that case I'd want each scope on a separate line.

I think if GDScript could get named optional arguments we'd be able to make both print and a potential dump function much better. print could have sep=" ", end="\n" as default like in Python so you can print and dump easily without options (or only the options you want to change) but a sep="" can be used for current behavior of print and dump could have similar options.

Of course, named optional arguments would be super useful all over the place.

**Comment by Calinou** [comment]
> Would dump also write the name of the variable?

I would really like to implement this, but unfortunately, I have no idea how to do it.
